wasm/src/main/java/com/dylibso/chicory/wasm/Module.java

```diff
@@ -34,35 +34,49 @@
 public class Module {
     private final HashMap<String, CustomSection> customSections;
 
-    private TypeSection typeSection = new TypeSection();
-    private ImportSection importSection = new ImportSection();
-    private FunctionSection functionSection = new FunctionSection();
-    private TableSection tableSection = new TableSection();
-    private MemorySection memorySection;
-    private GlobalSection globalSection = new GlobalSection();
-    private ExportSection exportSection = new ExportSection();
-    private StartSection startSection;
-    private ElementSection elementSection = new ElementSection();
-    private CodeSection codeSection = new CodeSection();
-    private DataSection dataSection = new DataSection();
-    private DataCountSection dataCountSection;
-
-    private final List<Integer> ignoredSections = new ArrayList();
-
-    public Module() {
-        this.customSections = new HashMap<>();
-    }
-
-    public void setTypeSection(TypeSection typeSection) {
+    private final TypeSection typeSection;
+    private final ImportSection importSection;
+    private final FunctionSection functionSection;
+    private final TableSection tableSection;
+    private final MemorySection memorySection;
+    private final GlobalSection globalSection;
+    private final ExportSection exportSection;
+    private final StartSection startSection;
+    private final ElementSection elementSection;
+    private final CodeSection codeSection;
+    private final DataSection dataSection;
+    private final DataCountSection dataCountSection;
+    private final List<Integer> ignoredSections;
+
+    Module(
+            TypeSection typeSection,
+            ImportSection importSection,
+            FunctionSection functionSection,
+            TableSection tableSection,
+            MemorySection memorySection,
+            GlobalSection globalSection,
+            ExportSection exportSection,
+            StartSection startSection,
+            ElementSection elementSection,
+            CodeSection codeSection,
+            DataSection dataSection,
+            DataCountSection dataCountSection,
+            HashMap<String, CustomSection> customSections,
+            List<Integer> ignoredSections) {
         this.typeSection = requireNonNull(typeSection);
-    }
-
-    public void setFunctionSection(FunctionSection functionSection) {
+        this.importSection = requireNonNull(importSection);
         this.functionSection = requireNonNull(functionSection);
-    }
-
-    public void setExportSection(ExportSection exportSection) {
+        this.tableSection = requireNonNull(tableSection);
+        this.memorySection = memorySection;
+        this.globalSection = requireNonNull(globalSection);
         this.exportSection = requireNonNull(exportSection);
+        this.startSection = startSection;
```

Comment:
As a general rule of thumb, I'm against the usage of `Optionals` as they are a leaky abstraction in Java(since they are nullable themselves), e.g. you can always do:
```java
Optional<Integer> x =null;
```

but in this case seams "self contained" enough to give it a spin.

Reply:
(done)

Reply:
I've never found that to be an issue in practice. That is the same for any other reference type which is expected to never be null. Once you start enforcing "no nulls" in your code base, these issues go away.

Reply:
The important feature of `Optional` is that you have to account for it being absent vs present. With a reference type, you need to rely on documentation, annotations, etc., to know if it can be null. With `Optional`, it's type-safe and part of the API.